src/arch/ucx/machine.C

```diff
@@ -372,99 +438,37 @@ static void UcxRxReqCompleted(void *request, ucs_status_t status,
-
-        // Register the source buffer and send back to destination to perform GET
-
-        NcpyOperationInfo *ncpyOpInfo = (NcpyOperationInfo *)(req->msgBuf);
-        UCX_LOG(4, "Got ncpy size %d (meta size %d)", ncpyOpInfo->srcSize, ncpyOpInfo->ncpyOpInfoSize);
-        resetNcpyOpInfoPointers(ncpyOpInfo);
-
-        UcxRdmaInfo *info = (UcxRdmaInfo *)(ncpyOpInfo->srcLayerInfo + CmiGetRdmaCommonInfoSize());
-
-        UcxMemMap(info,
-                  (void *)ncpyOpInfo->srcPtr,
-                  ncpyOpInfo->srcSize);
-
-        ncpyOpInfo->isSrcRegistered = 1;
-
-        ncpyOpInfo->freeMe = CMK_FREE_NCPYOPINFO; // It's a message, not a realy ncpy Obj
-        UCX_LOG(4, "Reset ncpy size %d (meta size %d)", ncpyOpInfo->destSize, ncpyOpInfo->ncpyOpInfoSize);
-
-        // send back to destination process to perform GET
-        UcxSendMsg(CmiNodeOf(ncpyOpInfo->destPe), ncpyOpInfo->destPe,
-                   ncpyOpInfo->ncpyOpInfoSize, (char*)ncpyOpInfo,
-                   UCX_RMA_TAG_GET, UcxRmaSendCompletedAndFree);
-
-    } else if (info->sender_tag & UCX_RMA_TAG_GET) {
-        NcpyOperationInfo *ncpyOpInfo = (NcpyOperationInfo *)(req->msgBuf);
-        resetNcpyOpInfoPointers(ncpyOpInfo);
-
-        ncpyOpInfo->freeMe = CMK_FREE_NCPYOPINFO; // It's a message, not a real ncpy Obj
-        UcxRmaOp(ncpyOpInfo, UCX_RMA_OP_GET);
-
-    } else if (info->sender_tag & UCX_RMA_TAG_DEREG_AND_ACK) {
-        NcpyOperationInfo *ncpyOpInfo = (NcpyOperationInfo *)(req->msgBuf);
-        resetNcpyOpInfoPointers(ncpyOpInfo);
-        ncpyOpInfo->freeMe = CMK_FREE_NCPYOPINFO;
-
-        if(CmiMyNode() == CmiNodeOf(ncpyOpInfo->srcPe)) { // source node
-            LrtsDeregisterMem(ncpyOpInfo->srcPtr,
-                              ncpyOpInfo->srcLayerInfo + CmiGetRdmaCommonInfoSize(),
-                              ncpyOpInfo->srcPe,
-                              ncpyOpInfo->srcRegMode);
-
-            ncpyOpInfo->isSrcRegistered = 0; // Set isSrcRegistered to 0 after de-registration
-
-            // Invoke source ack
-            if(ncpyOpInfo->opMode != CMK_BCAST_EM_API) {
-                ncpyOpInfo->opMode = CMK_EM_API_SRC_ACK_INVOKE;
-                CmiInvokeNcpyAck(ncpyOpInfo);
-            }
-
-        } else if(CmiMyNode() == CmiNodeOf(ncpyOpInfo->destPe)) { // destination node
-
-            LrtsDeregisterMem(ncpyOpInfo->destPtr,
-                              ncpyOpInfo->destLayerInfo + CmiGetRdmaCommonInfoSize(),
-                              ncpyOpInfo->destPe,
-                              ncpyOpInfo->destRegMode);
-
-            ncpyOpInfo->isDestRegistered = 0; // Set isDestRegistered to 0 after de-registration
-
-            // Invoke destination ack
-            ncpyOpInfo->opMode = CMK_EM_API_DEST_ACK_INVOKE;
-            CmiInvokeNcpyAck(ncpyOpInfo);
-
-        } else {
-            CmiAbort(" Cannot de-register on a different node than the source or destinaton");
-        }
-    }
-#endif
-
     if (req->msgBuf != NULL) {
```

Comment:
How is this a condition for the request not completing immediately?

Reply:
request is returned by `ucp_tag_recv_nb`. If receive is completed immediately, completion callback is invoked inside `ucp_tag_recv_nb` and `req->msgBuf` is not initialized yet. When it is not initialized it should be `NULL`, because we set it to `NULL` every time when return it to memory pool ( before invoke `ucp_request_free`).  In the callback we set `req->completed` to `1`. So, as soon as `ucp_tag_recv_nb` returns, we check `req->completed` and initialize `req->msgBuf` if `req->completed` is `0`. Thus, when completion callback is invoked later, `req->msgBuf`  is not `NULL`

Reply:
Okay, thanks for the clarification, that makes sense. 

So, since we need to set `req->msgBuf` to `NULL` before calling `UcxPostRxReq`, shouldn't we have to call `UCX_REQUEST_FREE` when we first pre-post the buffers in `LrtsInit`? (Inside the for loop on machine.C:417)

Reply:
Yes, it is needed. It is done in `UcxRequestInit`, which is one time initializer for all UCX requests

Reply:
Oh okay, understood. So, `UcxRequestInit` being passed as `cParams.request_init` makes `ucp_init` to call `UcxRequestInit` on each of the contexts? It's a little strange since the allocation of `UcxRequest`s happens in line 412. So, when exactly does `UcxRequestInit` get called on the contexts (for the first time)?

Reply:
It just initializes every request before its first use. According to ucx docs (from ucp.h file)
```
 700     /**
 701      * Pointer to a routine that is used for the request initialization.
 702      * This function will be called only on the very first time a request memory
 703      * is initialized, and may not be called again if a request is reused.
 704      * If a request should be reset before the next reuse, it can be done before
 705      * calling @ref ucp_request_free.

Reply:
Okay, makes sense. So, the UCX runtime takes care of it.